docs/internals.md

```diff
@@ -1,10 +1,679 @@
+
+Note: When using optionals, due to the macros involved, you can only use the `some(<value>)`/`none(<typedesc>)` syntax.
+
+### **Multiple Adders**
+In addition to passing properties to an adder, you can naturally also have multiple different adders. They just need different names.
+
+Let's look at a `CustomBox` widget with 2 `seq[Widget]` fields that you add to with different adders:
+
+```nim
+import owlkettle
+
+viewable CustomBox:
+  myChildren1: seq[Widget]
+  myChildren2: seq[Widget]
+
+  adder add: 
+    widget.hasMyChildren1 = true 
+    widget.valMyChildren1.add child
+  
+  adder add2:
+    widget.hasMyChildren2 = true
+    widget.valMyChildren2.add child
+
+method view(state: CustomBoxState): Widget =
+  gui: 
+    Box(orient = OrientY):
+      Box():
+        Label(text = "First Box")
+        for widget in state.myChildren1:
+          insert widget
+        
+      Box():
+        Label(text = "Second Box")
+        for widget in state.myChildren2:
+          insert widget
+
+## The App
+viewable App:
+  discard
+
+method view(state: AppState): Widget =
+  gui:
+    Window:
+      CustomBox():
+        Label(text = "I was passed in from the outside") # Uses "add"-adder implicitly by default 
+        Label(text = "Me too!") {.add.} # Uses "add"-adder explicitly
+        Label(text = "Me three!") {.add2.} # Uses "add2"-adder explicitly
+
+when isMainModule:
+  brew(gui(App()))
+```
+
+If no adder is specified, `Widget`s will always be added using the `add` adder. Otherwise the adder defined by the pragma annotation will be used.
+
+### **Hooks**
+Hooks are a concept introduced by owlkettle that allows you to execute code throughout a widget's lifecycle, or when an action on one of its fields occurs.
+
+Most hooks are defined only for Widgets, some are defined for both and `property` is only available as a hook for fields.
+
+The available hooks are:
+  - (W) beforeBuild     : Executed only once before the `WidgetState` is created.
+  - (WF) build           : Executed once after `WidgetState` is instantiated from `Widget`. Default values have not yet been applied and will overwrite any values set within this hook.
+  - (W) afterBuild      : Executed only once after the `WidgetState` was created. Is executed *after* all default values have been applied.
+  - (W) connectEvents   : Only relevant for renderables. Executed every time a callback is supposed to be attached to the underlying GTK widget. It defines how to do so.
+  - (W) disconnectEvents: Only relevant for renderables. Executed every time a callback is supposed to be removed from the underlying GTK widget. It defines how to do so.
+  - (WF) update         : Executed every time `WidgetState` is updated by `Widget`.
+  - (F) property        : Executed every time the hook-field changed its value during the update or build phases.
+  - (F) read            : Used in `Dialog`. Executes every time the state of the underlying GTK-Widget changes.
```

Comment:
Maybe make this a markdown table with the columns `Widget/Field` (TODO: better name?), `Name`, `Description`

Reply:
Done.
I also added the information whether a hook is useful for a viewable, a renderable or both, since certain hooks (beforeBuild, (dis)connectEvents) make little sense on a viewable.

Feel free to tell me whether that should be reverted or not